group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -0,0 +1,868 @@
+
+        if (ownedTopicPartitions != null || memberEpoch == 0 || assignmentUpdated) {
+            response.setAssignment(createResponseAssignment(member));
+        }
+
+        return new Result<>(records, response);
+    }
+
+    /**
+     * Handles leave request from a consumer group member.
+     * @param groupId       The group id from the request.
+     * @param memberId      The member id from the request.
+     *
+     * @return A Result containing the ConsumerGroupHeartbeat response and
+     *         a list of records to update the state machine.
+     */
+    private Result<ConsumerGroupHeartbeatResponseData> consumerGroupLeave(
+        String groupId,
+        String memberId
+    ) throws ApiException {
+        List<Record> records = new ArrayList<>();
+
+        ConsumerGroup group = getOrMaybeCreateConsumerGroup(groupId, false);
+        ConsumerGroupMember member = group.getOrMaybeCreateMember(memberId, false);
+
+        log.info("[GroupId " + groupId + "] Member " + memberId + " left the consumer group.");
+
+        // Write tombstones for the member. The order matters here.
+        records.add(newCurrentAssignmentTombstoneRecord(groupId, memberId));
+        records.add(newTargetAssignmentTombstoneRecord(groupId, memberId));
+        records.add(newMemberSubscriptionTombstoneRecord(groupId, memberId));
+
+        // We update the subscription metadata without the leaving member.
+        Map<String, TopicMetadata> subscriptionMetadata = group.computeSubscriptionMetadata(
+            memberId,
+            null,
+            topicsImage
+        );
+
+        if (!subscriptionMetadata.equals(group.subscriptionMetadata())) {
+            log.info("[GroupId " + groupId + "] Computed new subscription metadata: "
+                + subscriptionMetadata + ".");
+            records.add(newGroupSubscriptionMetadataRecord(groupId, subscriptionMetadata));
+        }
+
+        // We bump the group epoch.
+        int groupEpoch = group.groupEpoch() + 1;
+        records.add(newGroupEpochRecord(groupId, groupEpoch));
+
+        // We update the target assignment for the group and write it to
+        // the log.
+        String assignorName = group.preferredServerAssignor(
+            member.memberId(),
+            Optional.empty()
+        ).orElse(defaultAssignor.name());
+
+        try {
+            TargetAssignmentBuilder.TargetAssignmentResult assignmentResult =
+                new TargetAssignmentBuilder(groupId, groupEpoch, assignors.get(assignorName))
+                    .withMembers(group.members())
+                    .withSubscriptionMetadata(subscriptionMetadata)
+                    .removeMember(member.memberId())
+                    .build();
+
+            log.info("[GroupId " + groupId + "] Computed a new target assignment for epoch " + groupEpoch + ": "
+                + assignmentResult.targetAssignment() + ".");
+
+            records.addAll(assignmentResult.records());
```

Comment:
we don't perform a reconciliation phase on leave group because once the existing members heartbeat, they will the metadata manager will notice the bumped groupEpoch (L619) and so targetAssignmentEpoch < groupEpoch will trigger the reconciliation. is this correct?

why do we need to create target assignment records here then?

Reply:
That's fair. We could remove it.